nxos_l3_interfaces: fix states, add new minor attributes

chrisvanheuveln opened the pull request:
Depends-On: https://github.com/ansible/ansible/pull/63014

##### SUMMARY

- Add new attributes to `nxos_l3_interfaces` module:
  - `dot1q` :  (`encapsulation dot1q ##`)
  - `redirects`:  (`ip redirects`)
  - `unreachables`: (`ip unreachables`)

- Fixes for broken states merged/deleted/replaced/overridden)
  - All of the states had idempotence issues
  - I made some major changes to the way ipv4/ipv6 address changes occur, particularly with `replaced`.
    - The biggest problem was that the original code does a wholesale `no ip address` when there's a change to just one address, say changing a `tag` on a single secondary would remove the all addresses on the interface. This creates an unacceptable level of churn on the device for such a minor change.
    - Another issue that comes with address changes is that the changes may have to occur in sequence; e.g. primaries must be created first but can't be removed while secondaries exist.
    - I added helper methods for ipv4 & ipv6 to make the minimum number of changes while ignoring addresses that are idempotent.
    - `replaced` was also ignoring interfaces in the `want` list when they had no attributes defined
    - `merged` was ignoring tag-only changes

Additional issues were found during testing. I added extensive test cases to the unit test to cover a variety of scenarios. Please refer to the test cases to help understand the code changes that were made. Thanks.

Ref #37375.

##### ISSUE TYPE
- Bugfix Pull Request
- with new attrs

##### COMPONENT NAME
`nxos_l3_interfaces`

##### ADDITIONAL INFORMATION
Tested on N3K,N6K,N7K,N9K,N9Kv (internal testbeds: `n9k-108,n6k-77,camden-nx-1,n7k-j,dt-n9k5-1,n3k-173,n3k-105,vergreen-nx-1,greensboro-nx-1,hamilton-nx-1`

chrisvanheuveln committed:
(WIP) nxos_l3_interfaces: fix states, add new minor attributes

chrisvanheuveln committed:
sa cleanup

chrisvanheuveln committed:
more regression fixes

chrisvanheuveln committed:
test_8 for add'l code coverage

trishnaguha reviewed (commented):
The following test is failing https://github.com/ansible/ansible/blob/781442ac097a7a6e75630ef5268587671632bbda/test/integration/targets/nxos_l3_interfaces/tests/cli/merged.yaml#L56
```

TASK [nxos_l3_interfaces : assert] **************************************************************************************************************************************************************
task path: /home/trishna/workspace/ansible/test/results/.tmp/integration/nxos_l3_interfaces-9eR1kS-ÅÑŚÌβŁÈ/test/integration/targets/nxos_l3_interfaces/tests/cli/merged.yaml:56
<localhost> attempting to start connection
<localhost> using connection plugin network_cli
Found ansible-connection at path /usr/bin/ansible-connection
<localhost> found existing local domain socket, using it!
<localhost> Response received, triggered 'persistent_buffer_read_timeout' timer of 0.1 seconds
<localhost> Response received, triggered 'persistent_buffer_read_timeout' timer of 0.1 seconds
<localhost> Response received, triggered 'persistent_buffer_read_timeout' timer of 0.1 seconds
<localhost> updating play_context for connection
<localhost> 
<localhost> local domain socket path is /home/trishna/.ansible/pc/e6b47aa962
fatal: [localhost]: FAILED! => changed=false 
  assertion: result.after|length == (ansible_facts.network_resources.l3_interfaces|length - 1)
  evaluated_to: false
  msg: Assertion failed
```

Tested on N9K-9000v (Version 7.0(3)I7(3))

chrisvanheuveln commented:
> The following test is failing

I'm guessing that you don't use mgmt0 in your testbeds, which would throw off the count. I'll find a different way to test this or just remove the count from the assert.

chrisvanheuveln committed:
Fix regressions to handle mgmt w/o IP

chrisvanheuveln committed:
add 'no system default switchport' to regression setups

chrisvanheuveln committed:
add err msg to terminal_stderr_re so that cli_config will catch L2 failures

NilashishC commented:
recheck

NilashishC commented:
recheck

pabelanger commented on test/integration/targets/prepare_nxos_tests/tasks/main.yml:
```diff
@@ -35,7 +35,19 @@
+    when: item.interface is regex('port-channel|loopback|\.[0-9]+$')
+    ignore_errors: yes
+  no_log: True
+
```
So, for this. I'd like to see each test clean up after isn't self, rather then prepare. The reason for this, doing it here, isn't really efficient. Today, we likely spend on about 1.5 hours just doing prepare_nxos_tests things. In the case of zuul testing I am working on, we don't actually run this code but create the variables in the inventory file.

chrisvanheuveln replied:
@pabelanger I'll remove this. It was a convenient solution for us since we share our testbeds with more than just ansible testing and they're often left in a stale state.

pabelanger replied:
thanks! in our case, our nxos instances are ephemeral, so between different PRs, we are safe from it.  I believe, we can many work to update your test-bed to do this before running ansible-test, and update tests to try and clean up leaked data.

chrisvanheuveln replied:
It's been removed. I'll add this cleanup to our local CI infra, that should cover the majority of cases. Agreed, we should also keep an eye out for dirty test teardowns.

pabelanger replied:
Hmm, do you mind removing the whitespace here too? Trying to get zuul to work, and this is a merge conflict for us

pabelanger reviewed (changes requested):
I'd like to see us change prepare_nxos_tests here.

NilashishC commented:
recheck

NilashishC commented:
recheck

NilashishC commented:
recheck

NilashishC commented:
@pabelanger CI is not getting triggered. Could you have a look?

pabelanger commented:
tests are currently only run on https://github.com/ansible/ansible/pull/65442, so we need to use depends-on with it. until all tests are passing, we won't enable nxos on all PRs.

pabelanger commented on test/integration/targets/nxos_l3_interfaces/tasks/main.yaml:
```diff
@@ -1,2 +1,13 @@
 ---
+- set_fact: nxos_int4="{{ intdataraw[4].interface }}"
```
So, we are working to to bring nxos online for zuul, and don't have access to 4 interfaces. Can we refactor this to be nxos_int2 or nxos_int3?

NilashishC replied:
@chrisvanheuveln @mikewiebe Here's the failing run https://object-storage-ca-ymq-1.vexxhost.net/v1/a0b4156a37f9453eb4ec7db5422272df/ansible_42/65442/36707999cc13d951b048d574fa9c2e0c95eb1123/third-party-check/ansible-test-network-integration-nxos-cli-python36/a3e17de/controller/ara-report/result/f0a89e88-826a-4efe-8a03-0db65ee4622c/

chrisvanheuveln replied:
refactored to int3

pabelanger commented on test/integration/targets/nxos_l3_interfaces/tests/cli/deleted.yaml:
```diff
@@ -2,20 +2,32 @@
     msg: "Start nxos_l3_interfaces deleted integration tests connection={{ ansible_connection }}"
 
-- set_fact: test_int1="{{ nxos_int1 }}"
+- set_fact:
```
same comment about nxos_int4

pabelanger commented on test/integration/targets/nxos_l3_interfaces/tests/cli/merged.yaml:
```diff
@@ -2,24 +2,31 @@
     msg: "Start nxos_l3_interfaces merged integration tests connection={{ ansible_connection }}"
 
-- set_fact: test_int1="{{ nxos_int1 }}"
+- set_fact:
```
same

pabelanger commented on test/integration/targets/nxos_l3_interfaces/tests/cli/replaced.yaml:
```diff
@@ -2,20 +2,32 @@
     msg: "Start nxos_l3_interfaces replaced integration tests connection={{ ansible_connection }}"
 
-- set_fact: test_int1="{{ nxos_int1 }}"
+- set_fact:
```
same

chrisvanheuveln committed:
regression test change: /int4/int3/

NilashishC commented:
@pabelanger So the recent CI run for this target seems to be failing with:
https://object-storage-ca-ymq-1.vexxhost.net/v1/a0b4156a37f9453eb4ec7db5422272df/ansible_42/65442/36707999cc13d951b048d574fa9c2e0c95eb1123/third-party-check/ansible-test-network-integration-nxos-cli-python36/c9f1fbf/controller/ara-report/result/4a99dc3d-12e7-4b25-a121-d811427c6aef/

This is because all the `prepare_nxos_tests` tasks are being skipped. For example:

https://object-storage-ca-ymq-1.vexxhost.net/v1/a0b4156a37f9453eb4ec7db5422272df/ansible_42/65442/36707999cc13d951b048d574fa9c2e0c95eb1123/third-party-check/ansible-test-network-integration-nxos-cli-python36/c9f1fbf/controller/ara-report/result/3194870c-7e4f-486a-ad33-5eeb1001be52/

I can't seem to figure out why.

NilashishC commented:
@chrisvanheuveln @pabelanger In order to handle https://github.com/ansible/ansible/pull/64853/files#diff-b6090f5fdcd1fd85bc3670add7cefadaR4 we probably need to add https://github.com/ansible/ansible/blob/stable-2.9/test/integration/targets/prepare_nxos_tests/tasks/main.yml#L31-L38 in 
{{ target }}/tasks/main/yml.

pabelanger commented:
@NilashishC yes, in an effort to speed up job runs, we don't run prepare_nxos_tests any more, and add info into the inventory file directly. So if we need more info, we need to update ansible-zuul-jobs with the new data.

chrisvanheuveln commented:
@pabelanger > we need to update ansible-zuul-jobs with the new data.

Is this something I need to modify, or will Nilashish tackle that?

NilashishC commented:
@chrisvanheuveln I guess you can push a commit to this PR itself to add the [missing bits](https://github.com/ansible/ansible/blob/stable-2.9/test/integration/targets/prepare_nxos_tests/tasks/main.yml#L31-L38) only for nxos_l3_interfaces, coz right now, that is the only tests that needs to drop mgmt0 from facts list.

chrisvanheuveln commented:
Sorry I'm a bit confused by the ask. Paul said that ansible-zuul-jobs is the part that needs to be updated.

chrisvanheuveln committed:
Add default rsvd_intf_len for Zuul CI

chrisvanheuveln commented:
@NilashishC Thanks for the patch. I committed it with minor changes. Should be good to go now.

NilashishC commented:
recheck

NilashishC commented:
recheck

NilashishC commented:
CI for this target is green https://object-storage-ca-ymq-1.vexxhost.net/v1/a0b4156a37f9453eb4ec7db5422272df/ansible_53/64853/41fdca86a7c53652db37fd38f70e7eda8bfd0bea/third-party-check/ansible-test-network-integration-nxos-cli-python36/3f8ba84/controller/ara-report/reports/c27d7aa6-716d-42f1-a182-a71a21212b69.html

NilashishC commented:
@chrisvanheuveln It seems that the module is not removing the secondary IP address before negating the primary one and since we do not have an entry for `Must delete secondary address before deleting primary address` in our terminal error regex, the task run is not failing. 

Here's a quick playbook that I tested your changes with -https://gist.github.com/NilashishC/eea773491d39fe80812bcc84510a2606. The `replaced` task, sends `no ip address 192.0.2.1/24` and then configures the IPv6 address specified. However, the negate command for the IPv4 address doesn't work since there's a secondary address present, which needs to be cleared first. 

So effectively, the `replaced` operation (a) works like merged, i.e., just adds the IPv6 address (b) is not idempotent as the module always tries to send the IPv4 negate command.

IMO, we should add a scenario like this in the integration tests too.

NilashishC commented:
recheck

NilashishC commented:
recheck

NilashishC commented:
recheck

NilashishC commented:
recheck

NilashishC commented:
recheck

NilashishC commented:
recheck

NilashishC commented:
recheck

chrisvanheuveln committed:
Fix replaced-with-no-ipaddr and ip redirect issues

chrisvanheuveln commented:
>  It seems that the module is not removing the secondary IP address before negating the primary one and since we do not have an entry for `Must delete secondary address before deleting primary address` in our terminal error regex, the task run is not failing.

@NilashishC I fixed the issue with the ip address removal, which was fairly straightforward. 

But while testing I noticed that `ip redirects` has a behavior that we failed to address. The issue is that when `ip address ... secondary` is added to an interface, the device will automatically disable redirects (`no ip redirects`); when all secondaries are removed from an interface redirects is re-enabled automatically (`ip redirects`). 

Unfortunately, the redirect auto-enable behavior is inconsistent on legacy platforms such as N3K or N7K - I found that the device would often fail to re-enable redirects, and I was not able to find a consistent trigger for the failure. My solution was to explicitly re-enable redirects for those legacy platforms.

I also found that it was important to process the redirect commands after ipv4-related changes since adding/removing secondaries affects the redirects state, so I added a simple helper method to move redirect commands to the end of the cmds list.

I added another test to the UT for the redirect issue. I also updated the integration test to address the ip address scenario you discovered.

NilashishC commented:
@chrisvanheuveln Thanks a lot for this PR!

NilashishC commented:
bot_status

chrisvanheuveln commented:
@NilashishC Thanks for merging this. I know it was complicated. :-)
